go/vt/vtgate/healthcheck.go

```diff
@@ -0,0 +1,279 @@
+	var errfunc tabletconn.ErrFunc
+	var err error
+
+	// func to read one health check response, and notify downstream component
+	processResponse := func() error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case shr, ok := <-stream:
+			if !ok {
+				return errfunc()
+			}
+			if shr.Target == nil || shr.RealtimeStats == nil || shr.RealtimeStats.HealthError != "" {
+				return fmt.Errorf("health stats is not valid: %v", shr)
+			}
+
+			if hcc.target == nil {
+				// The first time we see response for the endpoint.
+				// Add the healthCheckConn to maps, and notify downstream component for endpoint going up.
+				hc.mu.Lock()
+				hc.addrToConns[addr] = hcc
+				hc.addEndPointToTarget(shr.Target, endPoint)
+				// TODO: notify downstream component for endpoint going up
+				hc.mu.Unlock()
+			} else if hcc.target.TabletType != shr.Target.TabletType {
+				// Update targetToConn when tablettype changes.
+				hc.mu.Lock()
+				hc.deleteEndPointFromTarget(hcc.target, endPoint)
+				hc.addEndPointToTarget(shr.Target, endPoint)
+				hc.mu.Unlock()
+			}
+			hcc.mu.Lock()
+			hcc.target = shr.Target
+			hcc.stats = shr.RealtimeStats
+			// TODO: notify downstream for tablettype and realtimestats
+			hcc.mu.Unlock()
+			return nil
+		}
+	}
+
+	// func to connect to endpoint and start streaming.
+	// it makes sure we have a chance to signal the first connection attempt regardless whether it succeeds.
+	connectFunc := func() error {
+		defer func() {
+			// signal that connection attempt is done for the first time.
+			if wg != nil {
+				wg.Done()
+				wg = nil
+			}
+		}()
+
+		conn, err = tabletconn.GetDialer()(ctx, endPoint, "" /*keyspace*/, "" /*shard*/, pbt.TabletType_RDONLY, hc.connTimeout)
+		if err != nil {
+			return err
+		}
+		stream, errfunc, err = conn.StreamHealth(ctx)
+		if err != nil {
+			conn.Close()
+			return err
+		}
+		err = processResponse()
+		if err != nil {
+			conn.Close()
+			return err
+		}
+		hcc.mu.Lock()
+		hcc.conn = conn
+		hcc.mu.Unlock()
+		return nil
+	}
+
+	// main loop to retry health check if it fails
+	for {
+		err = connectFunc()
+		if err != nil {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			log.Errorf("cannot connect to %+v: %v", endPoint, err)
+			time.Sleep(hc.retryDelay)
+			continue
+		}
+		for {
+			err = processResponse()
+			if err != nil {
+				conn.Close()
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+				log.Errorf("error when streaming tablet health from %+v: %v", endPoint, err)
+				time.Sleep(hc.retryDelay)
+				break
+			}
+		}
+	}
+
+}
+
+// onAddrUpdate handles endpoint up/down events.
+func (hc *HealthCheck) onAddrUpdate(updates <-chan *AddrUpdate) {
+	for update := range updates {
+		if update.IsUp {
+			go hc.checkConn(update.EndPoint, nil)
+			continue
+		}
+		addr := hc.endPointToString(update.EndPoint)
+		hc.mu.RLock()
+		hcc := hc.addrToConns[addr]
+		hc.mu.RUnlock()
+		hcc.cancelFunc()
```

Comment:
Should the hcc be removed from the maps too?

(As I said earlier, I'd rather have a addEndPoint / removeEndpoint than this channel, then it'd be clearer anyway).